storage/src/mmr/hasher.rs

```diff
@@ -1,82 +1,213 @@
 //! Decorator for a cryptographic hasher that implements the MMR-specific hashing logic.
 
+use crate::mmr::{
+    iterator::{leaf_num_to_pos, leaf_pos_to_num},
+    Error, Storage,
+};
 use commonware_cryptography::Hasher as CHasher;
+use std::future::Future;
 
-/// Hasher decorator the MMR uses for computing leaf, node and root hashes.
-pub(crate) struct Hasher<'a, H: CHasher> {
+/// A trait for computing the various digests of an MMR.
+pub trait Hasher<H: CHasher>: Send + Sync {
+    /// Computes the digest for a leaf given its position and the element it represents.
+    fn leaf_digest(
+        &mut self,
+        pos: u64,
+        element: &[u8],
+    ) -> impl Future<Output = Result<H::Digest, Error>> + Send;
+
+    /// Computes the digest for a node given its position and the digests of its children.
+    fn node_digest(&mut self, pos: u64, left: &H::Digest, right: &H::Digest) -> H::Digest;
+
+    /// Computes the root for an MMR given its size and an iterator over the digests of its peaks in
+    /// decreasing order of height.
+    fn root_digest<'b>(
+        &mut self,
+        pos: u64,
+        peak_digests: impl Iterator<Item = &'b H::Digest>,
+    ) -> H::Digest;
+
+    /// Compute the digest of a byte slice.
+    fn digest(&mut self, data: &[u8]) -> H::Digest;
+
+    /// Access the underlying commonware_cryptography hasher.
+    fn c_hasher(&mut self) -> &mut H;
+}
+
+/// The standard hasher to use with an MMR for computing leaf, node and root digests.
+pub struct Basic<'a, H: CHasher> {
     hasher: &'a mut H,
 }
 
-impl<'a, H: CHasher> Hasher<'a, H> {
-    pub(crate) fn new(hasher: &'a mut H) -> Self {
+impl<'a, H: CHasher> Basic<'a, H> {
```

Comment:
nit: I wonder if `Transparent` might be a better name for this?

Reply:
Yeah I struggled with this naming. Not sure I find Transparent much more descriptive though 🤷 . Let me see if I can come up with something better.

Reply:
As discussed, let's do `Standard`

Reply:
changed to "Standard" as discussed IRL